### ISSUE TITLE
Serialization exception cause createRequest never fired

### ISSUE THREAD

**Issue by ghost** [jakm/txmsgpackrpc#2]
Hi Jakub,
First, thank you for your lib.
I'm a newbie in twisted so excuse me if i'm wrong.
I encouter a problem with serialization exception which cause createRequest on client side never fired.
Could you tell me if there is a workaround in this specific case ?

A simple test : 

==server.py==

``` python
#!/usr/bin/env python
# -*- coding: utf-8 -*-
import os
import sys
from twisted.internet import reactor, defer
from twisted.internet.defer import inlineCallbacks
from twisted.internet.task import deferLater
from twisted.python import log
from txmsgpackrpc.server import MsgpackRPCServer

class DummyPopo():
    def __init__(self,name):
        self.name = "dummy %s"%name


class RpcServer(MsgpackRPCServer):
    @inlineCallbacks
    def remote_gatherresults(self):
        try:
            popos = [deferLater(reactor,1,lambda:DummyPopo(x)) for x in xrange(0,4)]
            res = yield  defer.gatherResults(popos)
            print(">> send res : %s"%res)
            defer.returnValue(res)
        except Exception,e:
            raise e

if __name__ == "__main__":
    sock_file = "/tmp/popo"
    if os.path.exists(sock_file):
        os.remove(sock_file)
    log.startLogging(sys.stdout)
    reactor.listenUNIX(sock_file, RpcServer().getStreamFactory())
    reactor.run()
```

== client.py ==

``` python
#!/usr/bin/env python
# -*- coding: utf-8 -*-
import sys
from twisted.internet import reactor
from twisted.internet.defer import inlineCallbacks
from twisted.python import log
from txmsgpackrpc.client import connect_UNIX

@inlineCallbacks
def remotecall_gatherresults():
    c=None
    try:
        c = yield  connect_UNIX("/tmp/popo")
        print "call remote method"
        r = yield c.createRequest("gatherresults")
        print "results : %s"%r
    except Exception,e:
        print e

if __name__ == "__main__":
    log.startLogging(sys.stdout)
    reactor.callWhenRunning(remotecall_gatherresults)
    reactor.run()
```

Thanks in advance

matboul

**Comment by jakm** [comment]
Hi Mathieu,
thank you for reporting. I've fixed it by commit 0b68ba4.

Data returned by remote method was packed to message that looks like this tuple:

``` python
(1, 1, None, [<__main__.DummyPopo instance at 0x7ff854ae6d88>, <__main__.DummyPopo instance at 0x7ff854ae6998>, <__main__.DummyPopo instance at 0x7ff854ae6a28>, <__main__.DummyPopo instance at 0x7ff854ae6950>]).
```

Message was passed to msgpack.Packer.pack method which raised TypeError because of unserializable object. TypeError was then catched and raised txmsgpackrpc.error.SerializationError. SerializationError was handled well by protocol, there was prepared message tuple with traceback and postponed to Packer.pack method. At this point process failed. The packer object contained parts of previous message and returned data were in invalid format. Client's unpacker object couldn't detect end of stream and stood in incorrect state.

Correct vs incorrect data:

``` python
"\x94\x01\x01\xda\x04\x0eTraceback: <type 'exceptions.TypeError'>..."
```

``` python
"\x94\x01\x01\xc0\x94\x94\x01\x01\xda\x04\x0eTraceback: <type 'exceptions.TypeError'>..."
```

Client's output after fix:

```
> $ python client.py
2015-05-27 16:44:35+0200 [-] Log opened.
2015-05-27 16:44:35+0200 [-] Starting factory <txmsgpackrpc.factory.MsgpackClientFactory instance at 0x7f349bc645f0>
2015-05-27 16:44:35+0200 [-] call remote method
2015-05-27 16:44:36+0200 [MsgpackStreamProtocol,client] Traceback: <type 'exceptions.TypeError'>: can't serialize <__main__.DummyPopo instance at 0x7f61339398c0>
2015-05-27 16:44:36+0200 [MsgpackStreamProtocol,client] /usr/lib/python2.7/dist-packages/twisted/internet/defer.py:1155:gotResult
2015-05-27 16:44:36+0200 [MsgpackStreamProtocol,client] /usr/lib/python2.7/dist-packages/twisted/internet/defer.py:1142:_inlineCallbacks
2015-05-27 16:44:36+0200 [MsgpackStreamProtocol,client] /usr/lib/python2.7/dist-packages/twisted/internet/defer.py:382:callback
2015-05-27 16:44:36+0200 [MsgpackStreamProtocol,client] /usr/lib/python2.7/dist-packages/twisted/internet/defer.py:490:_startRunCallbacks
2015-05-27 16:44:36+0200 [MsgpackStreamProtocol,client] --- <exception caught here> ---
2015-05-27 16:44:36+0200 [MsgpackStreamProtocol,client] /usr/lib/python2.7/dist-packages/twisted/internet/defer.py:577:_runCallbacks
2015-05-27 16:44:36+0200 [MsgpackStreamProtocol,client] /home/jakm/Vyvoj/public/txmsgpackrpc/txmsgpackrpc/protocol.py:234:respondCallback
2015-05-27 16:44:36+0200 [MsgpackStreamProtocol,client] /home/jakm/Vyvoj/public/txmsgpackrpc/txmsgpackrpc/protocol.py:255:writeMessage
2015-05-27 16:44:36+0200 [MsgpackStreamProtocol,client] _packer.pyx:224:msgpack._packer.Packer.pack (msgpack/_packer.cpp:224)
2015-05-27 16:44:36+0200 [MsgpackStreamProtocol,client] _packer.pyx:226:msgpack._packer.Packer.pack (msgpack/_packer.cpp:226)
2015-05-27 16:44:36+0200 [MsgpackStreamProtocol,client] _packer.pyx:214:msgpack._packer.Packer._pack (msgpack/_packer.cpp:214)
2015-05-27 16:44:36+0200 [MsgpackStreamProtocol,client] _packer.pyx:214:msgpack._packer.Packer._pack (msgpack/_packer.cpp:214)
2015-05-27 16:44:36+0200 [MsgpackStreamProtocol,client] _packer.pyx:221:msgpack._packer.Packer._pack (msgpack/_packer.cpp:221)
2015-05-27 16:44:36+0200 [MsgpackStreamProtocol,client]
2015-05-27 16:44:36+0200 [MsgpackStreamProtocol,client] <twisted.internet.unix.Connector instance at 0x7f349bc646c8> will retry in 2 seconds
2015-05-27 16:44:36+0200 [MsgpackStreamProtocol,client] Stopping factory <txmsgpackrpc.factory.MsgpackClientFactory instance at 0x7f349bc645f0>
2015-05-27 16:44:36+0200 [-] Main loop terminated.
```

**Comment by ghost** [comment]
Great !!
Many thanks for your job Jakub !
